internal/controller/dnsrecord_controller.go

```diff
@@ -96,11 +97,36 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	}
 	dnsRecord := previous.DeepCopy()
 
+	managedZone := &v1alpha1.ManagedZone{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dnsRecord.Spec.ManagedZoneRef.Name,
+			Namespace: dnsRecord.Namespace,
+		},
+	}
+	err = r.Get(ctx, client.ObjectKeyFromObject(managedZone), managedZone, &client.GetOptions{})
+	if err != nil {
+		reason := "ProviderError"
+		message := fmt.Sprintf("The DNS provider failed to ensure the record: %v", provider.SanitizeError(err))
+		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
+		return r.updateStatus(ctx, previous, dnsRecord, false, err)
+	}
+
+	// Ensure provider secret exists before continuing
+	err = r.Get(ctx, client.ObjectKey{Namespace: managedZone.Namespace, Name: managedZone.Spec.SecretRef.Name}, &v1.Secret{}, &client.GetOptions{})
+	if err != nil {
+		secretMissing.WithLabelValues(managedZone.Name, managedZone.Namespace, managedZone.Spec.SecretRef.Name).Set(1)
+		reason := "ProviderError"
+		message := fmt.Sprintf("The DNS provider failed to ensure the record: %v", provider.SanitizeError(err))
+		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
+		return r.updateStatus(ctx, previous, dnsRecord, false, err)
+	}
+	secretMissing.WithLabelValues(managedZone.Name, managedZone.Namespace, managedZone.Spec.SecretRef.Name).Set(0)
```

Comment:
Doesn't feel like this is where this should happen. We might want to check the status of the ManagedZone somewhere to make sure its ready, but i don't think dnsrecord should to be aware of secrets.

Reply:
yes ok, maybe it makes more sense to look for this in the managedzone controller? Seeing as that triggers on secret changes now it should see it immediately anyway.